### PR TITLE
Add Rights to Events

### LINKED ISSUE #869: Fix collaborator rights check for data traffic
#### Summary

Application/Gateway collaborator cannot observe events when having sufficient rights

#### Steps to Reproduce

1. Create a new application
2. Add a collaborator and assign all possible application rights:
```
"RIGHT_APPLICATION_DELETE"
"RIGHT_APPLICATION_DEVICES_READ"
"RIGHT_APPLICATION_DEVICES_READ_KEYS"
"RIGHT_APPLICATION_DEVICES_WRITE"
"RIGHT_APPLICATION_DEVICES_WRITE_KEYS"
"RIGHT_APPLICATION_INFO"
"RIGHT_APPLICATION_LINK"
"RIGHT_APPLICATION_SETTINGS_API_KEYS"
"RIGHT_APPLICATION_SETTINGS_BASIC"
"RIGHT_APPLICATION_SETTINGS_COLLABORATORS"
"RIGHT_APPLICATION_TRAFFIC_DOWN_WRITE"
"RIGHT_APPLICATION_TRAFFIC_READ"
"RIGHT_APPLICATION_TRAFFIC_UP_WRITE"
```
3. Login as the collaborator
4. Observe no events being shown for the application

**However, this works if the collaborator gets the `RIGHT_APPLICATION_ALL` right.**

#### What do you see now?

No events being shown for the collaborator with all necessary rights
```
{
   "error":{
      "grpc_code":7,
      "http_code":403,
      "message":"error:pkg/auth/rights:insufficient_application_rights (insufficient rights for application `new-app`)",
      "http_status":"Forbidden",
      "details":[
         {
            "@type":"type.googleapis.com/ttn.lorawan.v3.ErrorDetails",
            "namespace":"pkg/auth/rights",
            "name":"insufficient_application_rights",
            "message_format":"insufficient rights for application `{uid}`",
            "attributes":{
               "uid":"new-app"
            },
            "correlation_id":"37879009e0164eea8e30fed8af74197c"
         }
      ]
   }
}
```

#### What do you want to see instead?

No error and successfully opened events stream for the collaborator with the `RIGHT_{APPLICATION|GATEWAY}_TRAFFIC_READ` right

#### Environment

AS

#### How do you propose to implement this?

...

#### Can you do this yourself and submit a Pull Request?

no

### LINKED ISSUE #904: Send EndDevice events from NS/AS/JS EndDevice Registries
#### Summary

Just like we do in the Identity Server, we should send EndDevice update/delete events from the NS/AS/JS EndDevice registries.

#### Why do we need this?

- Consistency with the IS
- Integrations may want to take action if a certain field is updated.

#### What is already there? What do you see now?

In the IS:

https://github.com/TheThingsNetwork/lorawan-stack/blob/master/pkg/identityserver/end_device_registry.go#L31
https://github.com/TheThingsNetwork/lorawan-stack/blob/master/pkg/identityserver/end_device_registry.go#L112

#### What is missing? What do you want to see?

Similar events in the NS/AS/JS registries

#### How do you propose to implement this?

- We'll need to define new events (`end_device.update` and `end_device.delete` are already in use by the IS).
- The Set func of the `xxEndDeviceRegistryServer`s needs to emit an update event for the device identifiers with the applied field mask
- The Delete func needs to emit a delete event for the device identifiers

#### Can you do this yourself and submit a Pull Request?

Don't have time right now, so maybe someone else?

### PR DESCRIPTION
**Pull request by htdvisser**
#### Summary

This Pull Request adds a `visibility` field to our events, which will carry the rights that are required to see each event. 

This Pull Request closes #869 and closes #904.

#### Changes

- Added `visibility` to Event proto
- Use visibility in pkg/events
- Refactored rights checking in pkg/events/grpc
    - We now check that the caller has at least one right on the entities when the call starts
	- for this we're using the now exported errors in pkg/auth/rights
- Set visibility on existing events
- Fixed missing event in IS
- Fixed missing collaborator identifiers in collaborator events in IS
- Added missing events to NS/AS/JS device registry RPCs

#### Notes for Reviewers

@bafonins please do a full review of this pull request and verify that this indeed resolves your issue.

@johanstokking @rvolosatovs please check that I set the correct rights on `pkg/{network,application,join}server` and that I correctly emit events from the NS/AS/JS device registries. See individual commits for smaller diffs.

#### Release Notes

- The Events endpoint no longer requires the `_ALL` right on requested entities. All events now have explicit visibility rules.
- Added events to device registries in the Network Server, Application Server and Join Server.

### CODE REVIEW

**Comment by htdvisser** [comment]
@bafonins are you sure that there are actually events?

```
htdvisser % ttn-lw-cli app api-key create admin-app --name "traffic key" --right-application-traffic-read
 DEBUG Using access token (valid until 2:30PM)
 DEBUG ccResolverWrapper: sending update to cc: {[{localhost:8884 0  <nil>}] <nil>}
 DEBUG pickfirstBalancer: HandleSubConnStateChange: 0xc000187a80, CONNECTING
 DEBUG pickfirstBalancer: HandleSubConnStateChange: 0xc000187a80, READY
 DEBUG Finished unary call                      duration=31.819094ms grpc_method=CreateAPIKey grpc_service=ttn.lorawan.v3.ApplicationAccess namespace=grpc
  INFO API key ID: K2BV3CDJ6RLKENRP23QLE3QJPZQURVFNXSUSVSQ
  INFO API key value: NNSXS.K2BV3CDJ6RLKENRP23QLE3QJPZQURVFNXSUSVSQ.4EWWI3HJLGWPJXFBRHGRHH3GBX3A7H5PWKPWYEOYYDKL7QKM5YUA
  WARN The API key value will never be shown again
  WARN Make sure to copy it to a safe place
{
  "id": "K2BV3CDJ6RLKENRP23QLE3QJPZQURVFNXSUSVSQ",
  "key": "NNSXS.K2BV3CDJ6RLKENRP23QLE3QJPZQURVFNXSUSVSQ.4EWWI3HJLGWPJXFBRHGRHH3GBX3A7H5PWKPWYEOYYDKL7QKM5YUA",
  "name": "traffic key",
  "rights": [
    "RIGHT_APPLICATION_TRAFFIC_READ"
  ]
}

~/projects/lorawan-stack
htdvisser % access_token="NNSXS.K2BV3CDJ6RLKENRP23QLE3QJPZQURVFNXSUSVSQ.4EWWI3HJLGWPJXFBRHGRHH3GBX3A7H5PWKPWYEOYYDKL7QKM5YUA"

~/projects/lorawan-stack
htdvisser % curl -v -H "Authorization: Bearer $access_token" --data '{"identifiers":[{"application_ids":{"application_id": "admin-app"}}]}' http://localhost:1885/api/v3/events
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 1885 (#0)
> POST /api/v3/events HTTP/1.1
> Host: localhost:1885
> User-Agent: curl/7.54.0
> Accept: */*
> Authorization: Bearer NNSXS.K2BV3CDJ6RLKENRP23QLE3QJPZQURVFNXSUSVSQ.4EWWI3HJLGWPJXFBRHGRHH3GBX3A7H5PWKPWYEOYYDKL7QKM5YUA
> Content-Length: 69
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 69 out of 69 bytes
< HTTP/1.1 200 OK
< Access-Control-Allow-Origin: *
< Access-Control-Expose-Headers: Link, Date, Content-Length, X-Total-Count
< Content-Type: application/json
< Vary: Accept-Encoding
< Vary: Origin
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< X-Request-Id: 01DGF9PG0CEDEQQAH2NX1MYASV
< X-Xss-Protection: 1; mode=block
< Date: Tue, 23 Jul 2019 11:33:54 GMT
< Transfer-Encoding: chunked
<
{"result":{"name":"ns.up.merge_metadata","time":"2019-07-23T11:33:54.187870Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"data":{"@type":"type.googleapis.com/google.protobuf.Value","value":1},"correlation_ids":["gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"ns.up.data.forward","time":"2019-07-23T11:33:54.189317Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"correlation_ids":["gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"as.up.data.receive","time":"2019-07-23T11:33:54.188186Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"correlation_ids":["as:up:01DGF9QYCCVSSC11RTCMFBEAQJ","gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"as.up.data.forward","time":"2019-07-23T11:33:54.195629Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"data":{"@type":"type.googleapis.com/ttn.lorawan.v3.ApplicationUp","end_device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"},"correlation_ids":["as:up:01DGF9QYCCVSSC11RTCMFBEAQJ","gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"uplink_message":{"f_port":1,"f_cnt":2,"frm_payload":"q80=","rx_metadata":[{"gateway_ids":{"gateway_id":"admin-gtw"},"time":"2019-07-23T11:33:53.910191Z","timestamp":2437057967,"uplink_token":"ChcKFQoJYWRtaW4tZ3R3EggBAgMEBQYHCBCvm4qKCQ=="}],"settings":{"data_rate":{"lora":{"bandwidth":125000,"spreading_factor":12}},"coding_rate":"4/5","frequency":"868100000","timestamp":2437057967,"time":"2019-07-23T11:33:53.910191Z"}}},"correlation_ids":["as:up:01DGF9QYCCVSSC11RTCMFBEAQJ","gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"ns.mac.dev_status.request","time":"2019-07-23T11:33:54.198448Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"correlation_ids":["gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"ns.mac.new_channel.request","time":"2019-07-23T11:33:54.199123Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"data":{"@type":"type.googleapis.com/ttn.lorawan.v3.MACCommand.NewChannelReq","channel_index":3,"frequency":"867100000","max_data_rate_index":5},"correlation_ids":["gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"ns.mac.new_channel.request","time":"2019-07-23T11:33:54.199880Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"data":{"@type":"type.googleapis.com/ttn.lorawan.v3.MACCommand.NewChannelReq","channel_index":4,"frequency":"867300000","max_data_rate_index":5},"correlation_ids":["gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"ns.mac.new_channel.request","time":"2019-07-23T11:33:54.205228Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"data":{"@type":"type.googleapis.com/ttn.lorawan.v3.MACCommand.NewChannelReq","channel_index":5,"frequency":"867500000","max_data_rate_index":5},"correlation_ids":["gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"ns.mac.new_channel.request","time":"2019-07-23T11:33:54.205906Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"data":{"@type":"type.googleapis.com/ttn.lorawan.v3.MACCommand.NewChannelReq","channel_index":6,"frequency":"867700000","max_data_rate_index":5},"correlation_ids":["gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"ns.mac.new_channel.request","time":"2019-07-23T11:33:54.206472Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"data":{"@type":"type.googleapis.com/ttn.lorawan.v3.MACCommand.NewChannelReq","channel_index":7,"frequency":"867900000","max_data_rate_index":5},"correlation_ids":["gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
{"result":{"name":"ns.mac.rx_timing_setup.request","time":"2019-07-23T11:33:54.207103Z","identifiers":[{"device_ids":{"device_id":"admin-dev-abp","application_ids":{"application_id":"admin-app"},"dev_addr":"01020304"}}],"data":{"@type":"type.googleapis.com/ttn.lorawan.v3.MACCommand.RxTimingSetupReq","delay":5},"correlation_ids":["gs:conn:01DGF9QY4GXJRXSWVAW73PTT5J","gs:uplink:01DGF9QY5RAPZM5R98TRYPXNJG","ns:uplink:01DGF9QY5T4GNH898TMBDXYZ7A","rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01DGF9QY5TNA9VHDZHFH69QQWE","rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01DGF9QY42EV36B5FF0VY9M6GR"],"origin":"mcf","visibility":{"rights":["RIGHT_APPLICATION_TRAFFIC_READ"]}}}
```

File: api/events.proto [hunk 1]
```diff
@@ -33,6 +34,8 @@ message Event {
   repeated string correlation_ids = 5 [(gogoproto.customname) = "CorrelationIDs", (validate.rules).repeated.items.string.max_len = 100];
   string origin = 6;
   map<string,bytes> context = 7;
+  // The event will be visible to a caller that has any of these rights.
+  Rights visibility = 8;
```
Comment (rvolosatovs) [inline api/events.proto]: I think `visibility` can be a bit unclear in this case - how about e.g. `RequiredRights`?

Comment (htdvisser) [reply api/events.proto]: Yeah I was struggling with that too and changed it a couple of times myself.

- `Rights rights` - what rights?
- `Rights required_rights` - rights required for what?
- `Rights required_rights_for_read` - that's long
- `Rights visibility` - not perfect but not terrible either

**Comment by htdvisser** [comment]
@bafonins I also tried the curl without triggering events, and I'm not getting a timeout either.

**Comment by bafonins** [comment]
Right, seems I had the wrong stack configuration.
Just to make it clear, starting with this PR we always open the events stream for any collaborator, but filter the events based on the collaborator rights?

**Comment by bafonins** [comment]
Right, seems I had the wrong stack configuration.
Just to make it clear, starting with this PR we always open the events stream for any collaborator, but filter the events based on the collaborator rights?

File: pkg/gatewayserver/observability.go [hunk 2]
```diff
@@ -25,19 +25,48 @@ import (
...
+		"gs.up.forward", "forward uplink message",
+		ttnpb.RIGHT_GATEWAY_TRAFFIC_READ,
+	)
+	evtFailUp = events.Define(
+		"gs.up.fail", "fail to handle uplink message",
+		ttnpb.RIGHT_GATEWAY_TRAFFIC_READ,
+	)
+	evtSendDown = events.Define(
+		"gs.down.send", "send downlink message",
+		ttnpb.RIGHT_GATEWAY_TRAFFIC_READ,
```
Comment (johanstokking) [inline pkg/gatewayserver/observability.go]: Also `TRAFFIC_WRITE`

File: pkg/gatewayserver/observability.go [hunk 3]
```diff
@@ -25,19 +25,48 @@ import (
...
+	)
+	evtTxSuccessDown = events.Define(
+		"gs.down.tx.success", "transmit downlink message successful",
+		ttnpb.RIGHT_GATEWAY_TRAFFIC_READ,
```
Comment (johanstokking) [inline pkg/gatewayserver/observability.go]: Idem

File: pkg/gatewayserver/observability.go [hunk 4]
```diff
@@ -25,19 +25,48 @@ import (
...
+	)
+	evtTxFailureDown = events.Define(
+		"gs.down.tx.fail", "transmit downlink message failure",
+		ttnpb.RIGHT_GATEWAY_TRAFFIC_READ,
```
Comment (johanstokking) [inline pkg/gatewayserver/observability.go]: Idem.

This is for feedback on written downlink

File: pkg/joinserver/grpc_deviceregistry.go [hunk 5]
```diff
@@ -23,11 +23,27 @@ import (
...
+	"go.thethings.network/lorawan-stack/pkg/events"
 	"go.thethings.network/lorawan-stack/pkg/joinserver/provisioning"
 	"go.thethings.network/lorawan-stack/pkg/ttnpb"
 	"go.thethings.network/lorawan-stack/pkg/types"
 )
 
+var (
+	evtCreateEndDevice = events.Define(
+		"js.end_device.create", "create end device",
+		ttnpb.RIGHT_APPLICATION_DEVICES_READ,
```
Comment (rvolosatovs) [inline pkg/joinserver/grpc_deviceregistry.go]: Also `ttnpb.RIGHT_APPLICATION_DEVICES_WRITE`?

File: pkg/joinserver/grpc_deviceregistry.go [hunk 6]
```diff
@@ -23,11 +23,27 @@ import (
...
+	)
+	evtUpdateEndDevice = events.Define(
+		"js.end_device.update", "update end device",
+		ttnpb.RIGHT_APPLICATION_DEVICES_READ,
+	)
+	evtDeleteEndDevice = events.Define(
+		"js.end_device.delete", "delete end device",
+		ttnpb.RIGHT_APPLICATION_DEVICES_READ,
```
Comment (rvolosatovs) [inline pkg/joinserver/grpc_deviceregistry.go]: IDEM

File: pkg/joinserver/grpc_deviceregistry.go [hunk 7]
```diff
@@ -23,11 +23,27 @@ import (
...
 )
 
+var (
+	evtCreateEndDevice = events.Define(
+		"js.end_device.create", "create end device",
+		ttnpb.RIGHT_APPLICATION_DEVICES_READ,
+	)
+	evtUpdateEndDevice = events.Define(
+		"js.end_device.update", "update end device",
+		ttnpb.RIGHT_APPLICATION_DEVICES_READ,
```
Comment (rvolosatovs) [inline pkg/joinserver/grpc_deviceregistry.go]: IDEM

File: pkg/networkserver/grpc_deviceregistry.go [hunk 8]
```diff
@@ -21,10 +21,26 @@ import (
...
 	"go.thethings.network/lorawan-stack/pkg/crypto/cryptoutil"
+	"go.thethings.network/lorawan-stack/pkg/events"
 	"go.thethings.network/lorawan-stack/pkg/log"
 	"go.thethings.network/lorawan-stack/pkg/ttnpb"
 )
 
+var (
+	evtCreateEndDevice = events.Define(
+		"ns.end_device.create", "create end device",
+		ttnpb.RIGHT_APPLICATION_DEVICES_READ,
```
Comment (rvolosatovs) [inline pkg/networkserver/grpc_deviceregistry.go]: IDEM

File: pkg/networkserver/grpc_deviceregistry.go [hunk 9]
```diff
@@ -21,10 +21,26 @@ import (
...
+	)
+	evtUpdateEndDevice = events.Define(
+		"ns.end_device.update", "update end device",
+		ttnpb.RIGHT_APPLICATION_DEVICES_READ,
```
Comment (rvolosatovs) [inline pkg/networkserver/grpc_deviceregistry.go]: IDEM

File: pkg/networkserver/grpc_deviceregistry.go [hunk 10]
```diff
@@ -21,10 +21,26 @@ import (
...
+	)
+	evtDeleteEndDevice = events.Define(
+		"ns.end_device.delete", "delete end device",
+		ttnpb.RIGHT_APPLICATION_DEVICES_READ,
```
Comment (rvolosatovs) [inline pkg/networkserver/grpc_deviceregistry.go]: IDEM

File: pkg/joinserver/grpc_deviceregistry.go (hunk 5 above)
Comment (johanstokking) [reply pkg/joinserver/grpc_deviceregistry.go]: No, because write-only doesn't give you read access to devices created by others

**Pull request merged**